pkg/randchar/generator.go

```diff
@@ -1,54 +1,306 @@
+type Option func(r Rand) (Rand, error)
+
+// minimum defines a minimum number of characters that must be from a given character set.
+type minimum struct {
+	count   int
+	charset Charset
+}
+
+// Min ensures the generated slice contains at least n characters from the given character set.
+// When multiple Min options are given with the same character set, the biggest minimum takes precedence.
+func Min(n int, charset Charset) Option {
+	return func(r Rand) (Rand, error) {
+		if n < 1 {
+			return r, errors.New("minimum must be at least 1")
+		}
+
+		if len(charset) == 0 {
+			return r, errors.New("minimum character set cannot be empty")
+		}
+
+		if !charset.IsSubset((r.base)) {
+			return r, errors.New("minimum character set must be a subset of the base character set")
+		}
+
+		// Ensure the biggest minimum takes precedence when same charset minimum applies.
+		for i, minimum := range r.minima {
+			if minimum.charset.Equal(charset) {
+				if minimum.count >= n {
+					return r, nil
+				}
+
+				r.minLen += n - minimum.count
+				r.minima[i].count = n
+				return r, nil
+			}
+		}
+
+		r.minima = append(r.minima, minimum{
+			count:   n,
+			charset: charset,
+		})
+		r.minLen += n
+		return r, nil
+	}
+}
+
+// WithReader allows you to set the reader used as source of randomness.
+// Do not use this unless you know what you're doing. By default, the
+// source of randomness is set to crypto/rand.Reader.
+func WithReader(reader io.Reader) Option {
+	return func(r Rand) (Rand, error) {
+		r.reader = reader
+		return r, nil
+	}
+}
+
+// Generate returns a randomly generated slice of characters that meets the requirements of the reader.
+func (r Rand) Generate(n int) ([]byte, error) {
+	if n < r.minLen {
+		return nil, errors.New("n cannot be smaller than the minimum required length of the generator")
+	}
+
+	if n < 1 {
+		return []byte{}, nil
+	}
+
+	result := make([]byte, n)
+	i := 0
+	var err error
+	for _, min := range r.minima {
+		err = min.charset.rand(r.reader, result[i:i+min.count])
+		if err != nil {
+			return nil, err
+		}
+		i += min.count
+	}
+
+	err = r.base.rand(r.reader, result[i:])
+	if err != nil {
+		return nil, err
+	}
+
+	// Only shuffle if we've applied the minima.
+	// Otherwise, the result is already random and
+	// we can save time by skipping the shuffle step.
+	if i > 0 {
+		err = shuffle(r.reader, result)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// shuffle randomly shuffles elements of a byte slice, using the Durstenfeld shuffle algorithm.
+func shuffle(reader io.Reader, data []byte) error {
+	var randomIndex *big.Int
+	var j int64
+	var err error
+	for i := len(data) - 1; i > 0; i-- {
+		randomIndex, err = rand.Int(reader, big.NewInt(int64(i)))
+		if err != nil {
+			return err
+		}
+		j = randomIndex.Int64()
+		data[i], data[j] = data[j], data[i]
+	}
+
+	return nil
+}
+
+// Charset is a byte slice with a set of unique characters.
+type Charset []byte
```

Comment:
Yes we could, though it's a minor optimization for something that is called almost never (a few times on startup). Moreover, it keeps the logic for selecting a random character from the set more straightforward. Instead of selecting a random index from `map` keys (which has some undeterministic logic behind it) we select a random index from a fixed byte slice which is deterministic. In other words, it has less moving parts.

We can optimize the internal representation of the character set later IMO. Not something we'd need to dwell on now. We could make it a private type and only expose the `NewCharset` operation, but then we'd have to return an interface.

Reply:
I like your proposal of making the charset type private and exposing an interface. This makes the internal representation unexposed, which makes it possible to change it later without breaking changes. :+1: 

Note that we can't make any guarantees of the charset operations only being used on startup. They are public operations in a public package, so they can be used in any way.

I agree that the charsets should be optimized for selecting random characters. That's what they are meant to be used for.
What do you mean with undeterministic logic? Isn't (selecting) something random supposed to be undeterministic?

Reply:
This is more gut feeling than anything else, but I like to know where my randomness comes from. A loop over a slice is fixed and randomness comes from `crypto/rand`. A loop over map keys is performed in random order, which adds additional randomness to that coming from `crypto/rand`. That's why I prefer the slice type, you know what you're dealing with. 

You're right about us not being able to give guarantees about where it's used. This is meant as a library so creating an interface would actually help to minimize the exposure and keep internal flexibility for later. Will add that when I have time.

Reply:
Let's keep it a byte slice, but hide the internal type by making it private and returning an interface, like you proposed. :+1: